### PR TITLE
Daniel Cogan

### PR DESCRIPTION
**Pull request by danield9tqh**

### CODE REVIEW

File: .project [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+    <buildCommand>
+      <name>org.scala-ide.sdt.core.scalabuilder</name>
+    </buildCommand>
+  </buildSpec>
+  <natures>
+    <nature>org.scala-ide.sdt.core.scalanature</nature>
+    <nature>org.eclipse.jdt.core.javanature</nature>
+  </natures>
+  <linkedResources> </linkedResources>
+</projectDescription>
```
Comment (jkutsko) [inline .project:13]: Don't think these files needed to be committed, but no problem.

File: reflection.md [hunk 2]
```diff
@@ -1,18 +1,44 @@
...
+remembered the syntax for implicit conversions. 
+
+After I knew this, creating and implicit conversion from a string also seemed
+pretty easy if I just used the `Concat RegularExpression` object on `Literals`. 
+I wanted to use a foldl however on the string and that took me some time
+to get the right syntax. I kept getting a type error because my first argument
+to `foldl` was an `EPSILON` but overall it returned a `Concat`. I decided to 
+look at the sample solutions and found that if I specified both of them as 
+`RegularExpressions`, the types would check out.
+
```
Comment (jkutsko) [inline reflection.md:14]: Interesting. Yeah, I just did it recursively, and it worked out. Foldl seems more Scala-y though.

File: reflection.md [hunk 3]
```diff
@@ -1,18 +1,44 @@
...
+It seems pretty easy to define any arbitrary regex with this DSL. One thing that
+is difficult however is to define a range such as all digits or all numbers. One
+solution is to just define a val digit as is done in the code and then reuse 
+that variable throughout, but what if you want to match any number 0-7 or
+another variation? That would be a lot of needless definitions. 
+
+I think one thing that most regex parsers have is a range such as [0-5]+ would 
+match one or more numbers from 0-5. To implement this, I would probably define
+a function `-` that took in two ints or two characters and returned a Union of
+all the numbers in that range.
```
Comment (jkutsko) [inline reflection.md:43]: interesting, I like that idea.

File: src/main/scala/dsls/regex/RegularExpression.scala [hunk 4]
```diff
@@ -11,6 +11,27 @@ package dsls.regex
...
+  
+}
+
+object RegularExpression {
+    implicit def apply(literal: Char) = new Literal(literal)
+    implicit def apply(literal: String) = {
+        val literals: List[RegularExpression] = literal.toList.map(Literal)
+        val first: RegularExpression = EPSILON
+        literals.foldLeft(first)(Concat)
+    }
```
Comment (jkutsko) [inline src/main/scala/dsls/regex/RegularExpression.scala:34]: I like the (basically) one-liner using foldl a lot.

File: src/main/scala/dsls/regex/RegularExpression.scala [hunk 5]
```diff
@@ -11,6 +11,27 @@ package dsls.regex
 abstract class RegularExpression {
   /** returns true if the given string matches this regular expression */
   def matches(string: String) = RegexMatcher.matches(string, this)
+  def ||(other: RegularExpression) = new Union(this, other)
+  def ~(other: RegularExpression) = new Concat(this, other)
+  def <*> = new Star(this)
+  def <+> = new Concat(this, this <*>)
```
Comment (aputman) [inline src/main/scala/dsls/regex/RegularExpression.scala:17]: No documentation...

**Pull request closed without merging**